Added new projects around Batch Processing, Manual Ack and Custom Queue Names

gvensan opened the pull request:
[1) Added the following sample projects:

Manual Message Acknowledgement (cloud-stream-manual-ack)
Using SpEL for custom queue names (cloud-stream-custom-queue-names)
Batch Publisher (cloud-stream-batch-publish)
Batch Consumer (cloud-stream-batch-consume)

2) Renamed references to "cloud-streams" to "cloud-stream" in the POM and application config file for correctness!!

gvensan committed:
Sample to demonstrate custom `MessageConverter` to handle converting null payloads to custom POJOs

gvensan committed:
Updated spring boot version to 2.6.4 to be compatible with spring cloud version 2021.0.1 (bom 2.3.0)

gvensan committed:
- Updated version info of spring cloud, spring boot and spring cloud bom on all samples.
- Added appropriate comments to the null-payload-converter sample and updated methods to generate events with null payload (as spring doesn't allow that)

gvensan committed:
1) Added the following sample projects:

Manual Message Acknowledgement (cloud-stream-manual-ack)
Using SpEL for custom queue names (cloud-stream-custom-queue-names)
Batch Publisher (cloud-stream-batch-publish)
Batch Consumer (cloud-stream-batch-consume)

2) Renamed references to "cloud-streams" to "cloud-stream" in the POM and applcation config file for correctness!!

Mrc0113 commented on cloud-stream-custom-queue-names/src/main/java/com/solace/samples/spring/scs/GeneratedQueueNames.java:
```diff
@@ -0,0 +1,37 @@
+	}
+	
+	@Bean
+	public Consumer<Message<String>> uppercase() {
```
let's make sure we concentrate on the custom queue names here. I'd suggest remove the manual ACK stuff.

gvensan replied:
The only reason I included the ack and rejection action here is to demonstrate that the error queue (with custom name) receives the rejected events. I will remove this and just consume the events.

Mrc0113 commented on cloud-stream-custom-queue-names/src/main/java/com/solace/samples/spring/scs/GeneratedQueueNames.java:
```diff
@@ -0,0 +1,37 @@
+import org.slf4j.LoggerFactory;
+
+@SpringBootApplication
+public class GeneratedQueueNames {
```
Considering changing class name to CustomQueueNames

gvensan replied:
Yep, will do.

Mrc0113 commented on cloud-stream-custom-queue-names/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,48 @@
+              useGroupNameInQueueName: true
+              useDestinationEncodingInQueueName: false
+              queueNamePrefix: QNAME
+              queueNameExpression: >-
```
I suggest adding a comment with what the queue will actually look like.

gvensan replied:
Done

Mrc0113 commented on cloud-stream-custom-queue-names/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,48 @@
+                connectRetries: -1
+                reconnectRetries: -1
+      solace:
+        bindings:
```
Consider adding another binding with its own custom queue name that is just a literal to show how simple it is.

gvensan replied:
Added

Mrc0113 commented on cloud-stream-manual-ack/README.md:
```diff
@@ -0,0 +1,131 @@
+# Manual Message Acknowledgement
+
+In general, a received message is automatically acknowledged. This is true whether the message is consumed or results in an error. To manage the acknowledgement manually, the message handlers can disable auto-acknowledgment and manually invoke the acknowledgement callback. 
```
same language that needed to be changed in blog, maybe it just needs to be copied over?

Mrc0113 replied:
make other changes to this README if language changed in the blog as well.

gvensan replied:
Copied over the blog text into README (portions of it as applicable)

Mrc0113 commented on cloud-stream-manual-ack/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,38 @@
+          environment:
+            solace:
+              java:
+                host: tcp://localhost:55554
```
let's try to keep all the samples using `55555` for consistency.

Mrc0113 commented on cloud-stream-manual-ack/pom.xml:
```diff
@@ -0,0 +1,96 @@
+		</dependency>
+		<dependency>
+			<groupId>org.springframework.boot</groupId>
+			<artifactId>spring-boot-starter-web</artifactId>
```
I believe we can remove this requirement here. This will allow us to remove an extra dependency + remove the `server.port` parameter in the config.

Mrc0113 commented on cloud-stream-manual-ack/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,38 @@
+              autoBindErrorQueue: true
+              queueMaxMsgRedelivery: 2
+server:
+  port: 8096
```
remove this if not needed (if you can remove `spring-boot-starter-web` dependency. That shouldn't be required to show how to do manual acks

Mrc0113 commented on cloud-stream-sink-null-payload-converter/pom.xml:
```diff
@@ -0,0 +1,96 @@
+	<modelVersion>4.0.0</modelVersion>
+
+	<groupId>com.solace.samples.spring.scs</groupId>
+	<artifactId>cloud-stream-sink</artifactId>
```
update name to match sample `cloud-stream-sink-null-payload-converter`

Mrc0113 commented on cloud-stream-sink-null-payload-converter/pom.xml:
```diff
@@ -0,0 +1,96 @@
+	<artifactId>cloud-stream-sink</artifactId>
+	<version>0.0.2-SNAPSHOT</version>
+	<packaging>jar</packaging>
+	<name>cloud-stream-sink</name>
```
update name to match sample `cloud-stream-sink-null-payload-converter`

Mrc0113 commented on cloud-stream-sink-null-payload-converter/pom.xml:
```diff
@@ -0,0 +1,96 @@
+	<version>0.0.2-SNAPSHOT</version>
+	<packaging>jar</packaging>
+	<name>cloud-stream-sink</name>
+	<description>Sample sink for Spring Cloud Stream using Solace PubSub+ binder</description>
```
update description to match this sample

gvensan replied:
Updated the project!

Mrc0113 commented on cloud-stream-sink-null-payload-converter/src/main/resources/application.yml:
```diff
@@ -0,0 +1,38 @@
+    com.solacesystems.jcsmp: INFO
+    org.springframework: INFO
+    
+server:
```
remove if not needed

gvensan replied:
Sure, removed. In fact, removed from other application.yaml files as well.

Mrc0113 replied:
did the change get pushed? I'm still seeing the server.port for some reason.

Mrc0113 commented on cloud-stream-sink-null-payload-converter/src/main/resources/application.yml:
```diff
@@ -0,0 +1,38 @@
+          destination: TEMPS.Q
+          #The presence of "group" tells the binder to follow the "consumer group" pattern
+          group: SINK
+          consumer:
```
consider removing `consumer.concurrency` if not needed. keep things short and simple :)

gvensan replied:
Done

Mrc0113 commented on cloud-stream-sink-null-payload-converter/src/main/java/com/solace/samples/spring/scs/TemperatureSink.java:
```diff
@@ -0,0 +1,59 @@
+import com.solace.samples.spring.common.SensorReading;
+
+@SpringBootApplication
+public class TemperatureSink {
```
I think we need to create a unique class name here or this will conflict with what is in the `cloud-stream-sink` sample. no?

gvensan replied:
Done - renamed it to SensorReadingSink

Mrc0113 commented on cloud-stream-sink-null-payload-converter/src/test/java/com/solace/samples/spring/scs/TemperatureSinkTest.java:
```diff
@@ -0,0 +1,59 @@
+
+@ExtendWith(SpringExtension.class)
+@SpringBootTest
+public class TemperatureSinkTest {
```
I think we need to create a unique class name here or this will conflict with what is in the `cloud-stream-sink` sample. no?

gvensan replied:
Done

Mrc0113 reviewed (changes requested):
Overall it looks good, but I did make some comments requesting changes. 

Also, please add a sample showing batch consumption if you have it handy

gvensan committed:
- Renaming the class file to 'CustomQueueNames'
- Added an example for literal queueu name in queueNameExpression property
- Removed the ack code to keep it simple

gvensan committed:
- Adding initial version of batch-consumer project
- Updating the localhost:XXXX port to be more friendly for windows setup, changed from 55554 to 55555

gvensan committed:
Updated README with example for literal queue name example.

gvensan committed:
Update ManualAck project README to match the blog text.

gvensan committed:
- Updated null-payload handling project to match blog
- Updated application configuration file to remove the logger: settings (not used)

gvensan committed:
Renamed the class file to avoid collision with other projects.

Mrc0113 commented on cloud-stream-batch-consume/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,28 @@
+          content-type: 'application/octet-stream'
+          consumer:
+            batch-mode: true
+            useNativeDecoding: true
```
I'm honestly not sure what this does. is it required here? It also says if this is specified it doesn't use `content-type` which is specified above. 

Suggestion: just switch to using a message with a `String` payload to keep it simple. Plus it will work with teh batch publisher that uses messages with `String` payloads.

gvensan replied:
You are right - I did these to get it working because the data was coming in as byte[] by default from Try-Me.

Now, tried with just useNativeEncoing:true (no content-type), and changing the consumer function to accept List<String> - it just works.

Made the changes!

Mrc0113 commented on cloud-stream-batch-consume/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,28 @@
+        batchConsume-in-0:
+          destination: 'batch/consume/topic'
+          group: batch
+          content-type: 'application/octet-stream'
```
suggest removing.

gvensan replied:
I am assuming you are referring to the removal of content-type - the others have to stay, right!

Removed the content-type property (last conversation) as well.

Mrc0113 replied:
correct, I mean `content-type`

Mrc0113 commented on cloud-stream-batch-consume/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,28 @@
+                connectRetries: -1
+                reconnectRetries: -1
+server:
+  port: 8097
```
i believe you removed thie from the other samples

gvensan replied:
Did the removal on other yaml files as well.

Mrc0113 replied:
ty

Mrc0113 commented on cloud-stream-custom-queue-names/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,60 @@
+              provisionDurableQueue: true
+              addDestinationAsSubscriptionToQueue: true
+              useGroupNameInQueueName: true
+              useDestinationEncodingInQueueName: false
```
i believe this was deprecated and not necessary with the new queueNameExpression

gvensan replied:
Done

Mrc0113 commented on cloud-stream-custom-queue-names/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,60 @@
+            consumer:
+              provisionDurableQueue: true
+              addDestinationAsSubscriptionToQueue: true
+              useGroupNameInQueueName: true
```
i believe this was deprecated and not necessary with the new queueNameExpression

gvensan replied:
Yes, it is deprecated. Removed them.

Mrc0113 commented on cloud-stream-custom-queue-names/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,60 @@
+          uppercase-in-0:
+            consumer:
+              provisionDurableQueue: true
+              addDestinationAsSubscriptionToQueue: true
```
i believe this was deprecated and not necessary with the new queueNameExpression

gvensan replied:
Done

Mrc0113 commented on cloud-stream-custom-queue-names/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,60 @@
+        bindings:
+          uppercase-in-0:
+            consumer:
+              provisionDurableQueue: true
```
remove, as this is the default.

gvensan replied:
Done

Mrc0113 commented on cloud-stream-custom-queue-names/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,60 @@
+              addDestinationAsSubscriptionToQueue: true
+              useGroupNameInQueueName: true
+              useDestinationEncodingInQueueName: false
+              queueNamePrefix: QNAME
```
i believe this was deprecated and not necessary with the new queueNameExpression

gvensan replied:
Done

Mrc0113 commented on cloud-stream-custom-queue-names/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,60 @@
+              autoBindErrorQueue: true
+          lowercase-in-0:
+            consumer:
+              provisionDurableQueue: true
```
remove as this is the default

gvensan replied:
Done

Mrc0113 commented on cloud-stream-custom-queue-names/src/main/resources/application.yaml:
```diff
@@ -0,0 +1,60 @@
+              provisionDurableQueue: true
+              queueNameExpression: '''solace/just/a/literal/queuename'''
+server:
+  port: 8096
```
remove

gvensan replied:
Done

Mrc0113 reviewed (changes requested):
Please see some more comments I added. 
Minor updates to Batch Consumer. 
Also, I would say go ahead and remove the `server.port` and the web dependency from all of them since we aren't showing actuator usage.

gvensan committed:
Remove deprecated properties from the application configuration (yaml) file
